assignments/ganeodolu/src/components/TodoInput.tsx

```diff
@@ -0,0 +1,42 @@
+import { useState } from "react";
+import { useTodosActions } from '../context/TodosProvider';
+
+type HandleSubmitEvent = (e: React.SyntheticEvent<HTMLFormElement>) => void;
+
+type HandleChangeEvent = (e: React.ChangeEvent<HTMLInputElement>) => void;
+
+const TodoInput = () => {
+  const [content, setContent] = useState('');
+  const { add } = useTodosActions();
+
+  const handleInputChange: HandleChangeEvent = (e) => {
+    setContent(e.target.value);
+  }
```

Comment:
```typescript
  const handleInputSubmit: HandleSubmitEvent = e => {
    e.preventDefault();
    const todoInputValue = document.forms['todoInputContainer'].elements['todoInput'].value;
    if (todoInputValue.length === 0) return;
    add(todoInputValue);
    document.forms['todoInputContainer'].elements['todoInput'].value = '';
  };

  return (
    <header>
      <h1 className="header-text">Todos</h1>
      <form name="todoInputContainer" className="todo-input-container" onSubmit={handleInputSubmit}>
        <input type="text" name="todoInput" className="todo-input" placeholder="할일을 입력해주세요" />
      </form>
    </header>
  );
```

form으로 직접 접근은 처음 해봤는데 재미있네요. useState 관련 코드도 모두 생략가능하고요. 
다만 form 접근시 `Element implicitly has an 'any' type because index expression is not of type 'number'` 타입 오류가 나왔는데 `"suppressImplicitAnyIndexErrors": true,`로 제외했어요ㅎㅎㅎ

Reply:
```typescript
  const inputValueRef = useRef<HTMLInputElement>(null);
  const { add } = useTodosActions();

  const handleInputSubmit: HandleSubmitEvent = e => {
    e.preventDefault();
    if (inputValueRef.current) {
      const inputValue = inputValueRef.current.value;
      if (inputValue.length === 0) return;
      add(inputValue);
      inputValueRef.current.value = '';
    }
  };

  return (
    <header>
      <h1 className="header-text">Todos</h1>
      <form name="todoInputContainer" className="todo-input-container" onSubmit={handleInputSubmit}>
        <input
          ref={inputValueRef}
          type="text"
          name="todoInput"
          className="todo-input"
          placeholder="할일을 입력해주세요"
        />
      </form>
    </header>
  );

```
리액트 답게 useRef로도 구현해보았어요. 덕분에 코드가 점점 깔끔해지는 것 같아요 🟦👍🏼

Reply:
1. tsconfig를 직접 수정하면서 에러를 예외하는 것은 조금 아쉬운 부분입니다.

2. e 객체는 submit이 발생한 Form 객체가 들어있는데, 해당 타입이 event.target으로 추론되지 않을 거에요. 이런 DOM 컨트롤은 아래와 같이 체크해주시면 좋습니다.

  ```ts
  const target = event.target as HTMLFormElement;
  ```

3. 2번과 연계하여, document가 아니라 event.target.elements로 접근할 수 있겠죠?

4. 저 에러의 원인은 `[]`로 접근할 때 index가 number 타입이 아니기 때문에 발생합니다. 타입스크립트는 객체의 접근 요소가 프로퍼티 키인지, 인덱스 번호인지 알 수 없어요. 그래서 lib.dom.d.ts 등의 타입 선언 파일을 뒤져보면 DOM 객체가 가진 프로퍼티를 정의해두었는데 타입을 확인해보면서 작업하시는 것도 좋은 방법입니다.

5. 4번과 연계하여, `namedItem`이라는 키워드를 알려드릴게요. 찾아보시고 적용한 뒤에 tsconfig 설정을 지워보셔요!

Reply:
덕분에 오류의 원인을 정확히 알 수 있었어요. 파랑님 리뷰 바로 위와 같이 리액트 형식으로 바꿨기 때문에 tsconfig 설정은 지웠지만 다시 form 객체와 namedItem을 사용해서 아래와 같이 구현해보았습니다(타입 단언을 두번 사용한 것이 살짝 맘에 걸리네요). 😄 
```typescript
  const handleInputSubmit: HandleSubmitEvent = e => {
    e.preventDefault();
    const elementTarget = e.target as HTMLFormElement;
    const todoInputElement = elementTarget.elements.namedItem('todoInput') as HTMLInputElement;
    if (!todoInputElement) return;
    if (todoInputElement.value.length === 0) return;
    add(todoInputElement.value);
    todoInputElement.value = '';
  };
```